hothouse/ray_callbacks.pyx

```diff
@@ -0,0 +1,217 @@
+# cython: language=c++
+
+cimport numpy as np
+import numpy as np
+from pyembree.callback_handler cimport \
+    RayCollisionCallback, CALLBACK_TERMINATE, CALLBACK_CONTINUE
+from pyembree.rtcore_geometry cimport RTC_INVALID_GEOMETRY_ID
+from pyembree.rtcore_ray cimport RTCRay
+
+cdef class RayCollisionPrinter(RayCollisionCallback):
+    cdef int callback(self, RTCRay &ray):
+        print("Hi!", ray.geomID)
+        return CALLBACK_TERMINATE
+
+from cpython.mem cimport PyMem_Malloc, PyMem_Realloc, PyMem_Free
+
+cdef class RayCollisionMultiBounce(RayCollisionCallback):
+
+    cdef int nray
+    cdef int maxbounce
+    cdef float power_threshold
+    cdef int iray
+    cdef float ipower
+    cdef public np.int32_t[:] nbounce
+    cdef public np.int32_t[:,:] primID
+    cdef public np.int32_t[:,:] geomID
+    cdef public np.int32_t[:,:] instID
+    cdef public np.float32_t[:,:] tfars
+    cdef public np.float32_t[:,:,:] Ng
+    cdef public np.float32_t[:,:,:] ray_dir
+    cdef public np.float32_t[:,:] power
+    cdef int nq
+    cdef RTCRay* ray_queue
+    cdef float* power_queue
+    cdef float** transmittance
+    cdef float** reflectance
+
+    def __cinit__(self, int nray, int maxbounce,
+                  list transmittance, list reflectance):
+        self.nray = nray
+        self.maxbounce = maxbounce
+        # TODO: Pass this in
+        self.power_threshold = 0.001
+        self.iray = 0
+        self.ipower = 1.0
+        self.nbounce = np.zeros(nray, dtype="int32")
+        self.primID = -1 * np.ones((nray, maxbounce), dtype="int32")
+        self.geomID = -1 * np.ones((nray, maxbounce), dtype="int32")
+        self.instID = -1 * np.ones((nray, maxbounce), dtype="int32")
+        self.tfars = np.empty((nray, maxbounce), dtype="float32")
+        self.Ng = np.empty((nray, maxbounce, 3), dtype="float32")
+        self.ray_dir = np.empty((nray, maxbounce, 3), dtype="float32")
+        self.power = np.zeros((nray, maxbounce), dtype="float32")
+        self.nq = 0
+        self.ray_queue = <RTCRay*>PyMem_Malloc(self.maxbounce * sizeof(RTCRay))
```

Comment:
We're using C++ anyway -- would it make sense to make this a vector?

Reply:
If we arn't reallocating then this can just be an array on stack. I had it dynamic originally and was adjusting the size as needed, but then realized that that required a dynamic location to store the rays as well for use outside the callback. Since this presents memory issues for low thresholds, I figured it would be easier to just assume a max number of additional rays (maxbounce). This has some issues since it might truncate significant rays in the current form which always follows the reflected ray first. I think it should probably follow the queued ray with the largest power first (implemented in the pop_ray method), but maybe there is a better way to do this.